File: code/programs/elixir/unix-tools/test/basename_test.exs
defmodule BasenameTest do
  @moduledoc """
  Tests for the basename tool.

  ## What These Tests Verify

  These tests cover:
  1. CLI Builder integration (parsing flags, --help, --version, errors).
  2. Business logic (compute_basename with various paths and suffixes).
  3. Edge cases (trailing slashes, all-slashes path, suffix equals name).
  """

  use ExUnit.Case, async: true

  alias CodingAdventures.CliBuilder.Parser
  alias CodingAdventures.CliBuilder.{ParseResult, HelpResult, VersionResult, ParseErrors}

  # ---------------------------------------------------------------------------
  # Helper: locate the spec file
  # ---------------------------------------------------------------------------

  @spec_path Path.join([__DIR__, "..", "basename.json"]) |> Path.expand()

  # ---------------------------------------------------------------------------
  # Helper: parse argv against the basename spec
  # ---------------------------------------------------------------------------

  defp parse_argv(argv) do
    Parser.parse(@spec_path, argv)
  end

  # ---------------------------------------------------------------------------
  # Test: Default behavior
  # ---------------------------------------------------------------------------

  describe "default behavior" do
    test "single name argument is captured" do
      {:ok, %ParseResult{arguments: arguments}} = parse_argv(["basename", "/usr/bin/sort"])
      assert arguments["name"] == ["/usr/bin/sort"]
    end

    test "multiple name arguments are captured" do
      {:ok, %ParseResult{arguments: arguments}} =
        parse_argv(["basename", "-a", "/usr/bin/sort", "/usr/bin/ls"])

      assert arguments["name"] == ["/usr/bin/sort", "/usr/bin/ls"]
    end
  end

  # ---------------------------------------------------------------------------
  # Test: Flags
  # ---------------------------------------------------------------------------

  describe "flags" do
    test "-a sets multiple to true" do
      {:ok, %ParseResult{flags: flags}} = parse_argv(["basename", "-a", "file"])
      assert flags["multiple"] == true
    end

    test "-s sets suffix" do
      {:ok, %ParseResult{flags: flags}} = parse_argv(["basename", "-s", ".h", "file.h"])
      assert flags["suffix"] == ".h"
    end

    test "-z sets zero to true" do
      {:ok, %ParseResult{flags: flags}} = parse_argv(["basename", "-z", "file"])
      assert flags["zero"] == true
    end
  end

  # ---------------------------------------------------------------------------
  # Test: --help and --version
  # ---------------------------------------------------------------------------

  describe "--help flag" do
    test "returns {:ok, %HelpResult{}}" do
      assert {:ok, %HelpResult{}} = parse_argv(["basename", "--help"])
    end

    test "help text contains program name" do
      {:ok, %HelpResult{text: text}} = parse_argv(["basename", "--help"])
      assert text =~ "basename"
    end
  end

  describe "--version flag" do
    test "returns {:ok, %VersionResult{}}" do
      assert {:ok, %VersionResult{}} = parse_argv(["basename", "--version"])
    end

    test "version string is 1.0.0" do
      {:ok, %VersionResult{version: version}} = parse_argv(["basename", "--version"])
      assert version == "1.0.0"
    end
  end

  # ---------------------------------------------------------------------------
  # Test: Unknown flags
  # ---------------------------------------------------------------------------

  describe "unknown flags" do
    test "unknown long flag returns error" do
      assert {:error, %ParseErrors{}} = parse_argv(["basename", "--unknown", "file"])
    end
  end

  # ---------------------------------------------------------------------------
  # Test: Business logic - compute_basename
  # ---------------------------------------------------------------------------

  describe "compute_basename/2" do
    test "strips directory from absolute path" do
      assert UnixTools.BasenameTool.compute_basename("/usr/bin/sort") == "sort"
    end

    test "strips directory from relative path" do
      assert UnixTools.BasenameTool.compute_basename("include/stdio.h") == "stdio.h"
    end

    test "removes suffix when specified" do
      assert UnixTools.BasenameTool.compute_basename("stdio.h", ".h") == "stdio"
    end

    test "does not remove suffix when name equals suffix" do
      assert UnixTools.BasenameTool.compute_basename(".h", ".h") == ".h"
    end

    test "handles trailing slashes" do
      assert UnixTools.BasenameTool.compute_basename("/usr/bin/") == "bin"
    end

    test "handles all-slashes path" do
      assert UnixTools.BasenameTool.compute_basename("///") == "/"
    end

    test "handles root path" do
      assert UnixTools.BasenameTool.compute_basename("/") == "/"
    end

    test "handles plain filename" do
      assert UnixTools.BasenameTool.compute_basename("hello.txt") == "hello.txt"
    end

    test "strips suffix from path with directory" do
      assert UnixTools.BasenameTool.compute_basename("/usr/include/stdio.h", ".h") == "stdio"
    end

    test "nil suffix is ignored" do
      assert UnixTools.BasenameTool.compute_basename("/usr/bin/sort", nil) == "sort"
    end
  end
end

File: code/programs/rust/unix-tools/tests/basename_test.rs
//! # Integration Tests for basename
//!
//! These tests exercise both the CLI Builder integration and the basename
//! business logic. We test directory stripping, suffix removal, and
//! edge cases like root paths and empty strings.

use cli_builder::{load_spec_from_file, Parser, ParserOutput};
use unix_tools::basename_tool::strip_basename;
use std::path::PathBuf;

// ---------------------------------------------------------------------------
// Helper: locate the spec file
// ---------------------------------------------------------------------------

fn spec_path() -> String {
    let manifest_dir = env!("CARGO_MANIFEST_DIR");
    let path = PathBuf::from(manifest_dir).join("basename.json");
    path.to_string_lossy().into_owned()
}

fn make_parser() -> Parser {
    let spec = load_spec_from_file(&spec_path()).expect("failed to load basename.json");
    Parser::new(spec)
}

fn parse_argv(argv: &[&str]) -> ParserOutput {
    let parser = make_parser();
    let args: Vec<String> = argv.iter().map(|s| s.to_string()).collect();
    parser.parse(&args).expect("parse failed")
}

// ---------------------------------------------------------------------------
// Test: Spec loads successfully
// ---------------------------------------------------------------------------

#[cfg(test)]
mod spec_loading {
    use super::*;

    #[test]
    fn spec_loads() {
        let spec = load_spec_from_file(&spec_path());
        assert!(spec.is_ok(), "basename.json should load successfully");
    }
}

// ---------------------------------------------------------------------------
// Test: Flag parsing
// ---------------------------------------------------------------------------

#[cfg(test)]
mod flag_parsing {
    use super::*;

    #[test]
    fn with_name_arg() {
        match parse_argv(&["basename", "/usr/bin/sort"]) {
            ParserOutput::Parse(_) => {}
            other => panic!("expected Parse, got {:?}", other),
        }
    }

    #[test]
    fn multiple_flag() {
        match parse_argv(&["basename", "-a", "file1", "file2"]) {
            ParserOutput::Parse(result) => {
                assert_eq!(
                    result.flags.get("multiple"),
                    Some(&serde_json::json!(true))
                );
            }
            _ => panic!("expected Parse"),
        }
    }

    #[test]
    fn suffix_flag() {
        match parse_argv(&["basename", "-s", ".txt", "file.txt"]) {
            ParserOutput::Parse(result) => {
                assert_eq!(
                    result.flags.get("suffix").and_then(|v| v.as_str()),
                    Some(".txt")
                );
            }
            _ => panic!("expected Parse"),
        }
    }

    #[test]
    fn help_returns_help() {
        match parse_argv(&["basename", "--help"]) {
            ParserOutput::Help(_) => {}
            other => panic!("expected Help, got {:?}", other),
        }
    }

    #[test]
    fn version_returns_version() {
        match parse_argv(&["basename", "--version"]) {
            ParserOutput::Version(v) => {
                assert_eq!(v.version, "1.0.0");
            }
            _ => panic!("expected Version"),
        }
    }
}

// ---------------------------------------------------------------------------
// Test: Business logic — strip_basename
// ---------------------------------------------------------------------------

#[cfg(test)]
mod business_logic {
    use super::*;

    #[test]
    fn simple_path() {
        assert_eq!(strip_basename("/usr/bin/sort", None), "sort");
    }

    #[test]
    fn with_suffix() {
        assert_eq!(strip_basename("program.sh", Some(".sh")), "program");
    }

    #[test]
    fn suffix_not_matching() {
        assert_eq!(strip_basename("program.sh", Some(".txt")), "program.sh");
    }

    #[test]
    fn root_path() {
        assert_eq!(strip_basename("/", None), "/");
    }

    #[test]
    fn trailing_slash() {
        assert_eq!(strip_basename("/usr/bin/", None), "bin");
    }

    #[test]
    fn no_directory() {
        assert_eq!(strip_basename("hello", None), "hello");
    }

    #[test]
    fn empty_string() {
        assert_eq!(strip_basename("", None), "");
    }

    #[test]
    fn deep_path_with_suffix() {
        assert_eq!(
            strip_basename("/home/user/docs/report.pdf", Some(".pdf")),
            "report"
        );
    }
}
